### PR TITLE
Use `@babel/core` for AST bundle processing and disallow HTML comments

### PR DESCRIPTION
**Pull request by Mrtenz**
This changes how we use Babel for processing bundles. Rather than using the low-level `@babel/generator`, `@babel/parser` and `@babel/traverse` packages, we now use `@babel/core` instead. The bundle processing is registered as a plugin on Babel.

This has the advantage that we can easily check if code contains HTML comments (`<!--`, `-->`), by checking `path.toSource()`, which is not supported when using `@babel/traverse` directly. It also simplifies the code a little bit.

### CODE REVIEW

File: packages/utils/src/ast.ts [hunk 1]
```diff
@@ -389,7 +323,43 @@ export function postProcessAST(ast: Node): Node {
...
+
+      // Throw an error if HTML comments are used as a binary expression.
+      if (source.includes('<!--') || source.includes('-->')) {
+        throw new Error('HTML comments (`<!--` and `-->`) are not allowed.');
+      }
+    },
+  };
+
+  try {
+    const file = transformSync(code, {
```
Comment (FrederikBolding) [inline packages/utils/src/ast.ts:338]: Is this preferred over using the parser and generator directly?

File: packages/utils/src/ast.ts [hunk 2]
```diff
@@ -199,34 +142,25 @@ function getRawTemplateValue(value: string) {
...
-        comments.forEach((comment) => {
-          comment.value = comment.value
-            .replace(new RegExp(`<!${'--'}`, 'gu'), '< !--')
-            .replace(new RegExp(`${'--'}>`, 'gu'), '-- >')
-            .replace(/import(\(.*\))/gu, 'import\\$1');
-        });
-      });
-    },
-
+export function postProcessAST(code: string, attachComments = true): string {
```
Comment (FrederikBolding) [inline packages/utils/src/ast.ts:149]: Perhaps this should be renamed if it doesn't take an AST anymore.

File: packages/utils/src/ast.ts [hunk 3]
```diff
@@ -27,59 +18,11 @@ const TOKEN_REGEX = /(<!)(--)|(--)(>)|(import)(\(.*?\))/gu;
...
-  return generate(ast, {}, code).code;
-}
-
-const evalWrapper = template.smart(`
+const evalWrapper = template.statement(`
   (1, REF)(ARGS)
 `);
 
-const objectEvalWrapper = template.smart(`
+const objectEvalWrapper = template.statement(`
```
Comment (FrederikBolding) [inline packages/utils/src/ast.ts:25]: ? 🤔

File: packages/utils/src/ast.test.ts [hunk 4]
```diff
@@ -253,14 +182,21 @@ describe('postProcessAST', () => {
...
-
-    expect(getCode(processedCode)).toMatchInlineSnapshot(
+    const processedCode = postProcessAST(code);
+    expect(processedCode).toMatchInlineSnapshot(
       `"const foo = \`\${\\"<!\\"}\${\\"--\\"} \\\\\` \${foo} \\\\\` \${\\"--\\"}\${\\">\\"}\`;"`,
     );
   });
 
+  it.each(['const a = b <!-- c;', 'const a = b --> c;'])(
+    'throws an error when HTML comment tokens are used',
```
Comment (FrederikBolding) [inline packages/utils/src/ast.test.ts:192]:
```suggestion
    'throws an error when HTML comment tokens are used as operators',
```

File: packages/utils/src/ast.ts [hunk 5]
```diff
@@ -389,7 +323,43 @@ export function postProcessAST(ast: Node): Node {
...
     },
-  });
 
-  return ast;
+    BinaryExpression(path) {
+      const source = path.getSource();
+
+      // Throw an error if HTML comments are used as a binary expression.
+      if (source.includes('<!--') || source.includes('-->')) {
+        throw new Error('HTML comments (`<!--` and `-->`) are not allowed.');
```
Comment (FrederikBolding) [inline packages/utils/src/ast.ts:332]: Could we elaborate on this error saying that they aren't allowed to be used as operators?

Comment (FrederikBolding) [reply packages/utils/src/ast.ts:332]: Perhaps with a link to explain why or just a blurb that explains why

File: packages/utils/src/ast.ts (hunk 1 above)
Comment (Mrtenz) [reply packages/utils/src/ast.ts:338]: Yes, it's explained in the PR description why.

File: packages/utils/src/ast.ts (hunk 2 above)
Comment (Mrtenz) [reply packages/utils/src/ast.ts:149]: Will merge this with the other `postProcess` function.

File: packages/utils/src/ast.ts (hunk 1 above)
Comment (FrederikBolding) [reply packages/utils/src/ast.ts:338]: Yeah, I just meant in general, does Babel prefer that people do this over using the underlying APIs?

File: packages/utils/src/post-process.test.ts [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import { postProcessBundle } from './post-process';
 
-describe('postProcessBundle', () => {
+describe('postProcessAST', () => {
```
Comment (FrederikBolding) [inline packages/utils/src/post-process.test.ts:3]:
```suggestion
describe('postProcessBundle', () => {
```

File: packages/utils/src/post-process.ts [hunk 7]
```diff
@@ -1,40 +1,382 @@
...
+        );
+      }
+    },
+  };
+
+  try {
+    const file = transformSync(code, {
+      // Prevent Babel from searching for a config file.
+      configFile: false,
+      comments: !stripComments,
```
Comment (FrederikBolding) [inline packages/utils/src/post-process.ts:358]: Do we not want `parserOpts: { attachComment: !stripComments }`? We don't need to parse them if we are throwing them out right?

Comment (Mrtenz) [reply packages/utils/src/post-process.ts:358]: There's no `attachComment` in `parserOpts` 😕

**Review by FrederikBolding — APPROVED** [review]

**Pull request merged**